Add TestGrid Documentation and Linting (#10)
CSPI-QE/ocp-ci-docs: docs/OCP_CI_Tutorials/Scenarios/Reporting_Guide.md

# Reporting Guide<!-- omit from toc -->

## Table of Contents<!-- omit from toc -->

- [Introduction](#introduction)
- [TestGrid](#testgrid)
  - [TestGrid Configuration Example and Description](#testgrid-configuration-example-and-description)
  - [How to Create a New Dashboard](#how-to-create-a-new-dashboard)
  - [How to Add a Job to a Dashboard](#how-to-add-a-job-to-a-dashboard)
  - [Pull Request Process](#pull-request-process)
- [Report Portal](#report-portal)
- [Slack](#slack)

## Introduction

> Because OpenShift CI does not have reporting built into it's logic, we must configure and maintain our reporting step. This process may change as we evolve our processes in OpenShift CI, but we have tried to make this step as easy as possible. Please see each section below to see how to report your scenarios to each of the platforms.

Reviewer: ```suggestion
Because OpenShift CI does not have reporting built into its logic, we must configure and maintain our reporting step. This process may change as we evolve our processes in OpenShift CI, but we have tried to make this step as easy as possible. Please see each section below to see how to report your scenarios to each of the platforms.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> ## TestGrid

Reviewer 2: I was doing a little more investigation into test grid regarding dashboard creations. There are a couple of current ways dashboards get created today.

1. Like you have described here, open PR's into `kubernetes/test-infra` to add the dashboards configuration and then following PR's for each job to be included
2. Have the test grid config generator tool run on a daily basis and automatically generate the dashboard configs/open PR's into kuberentes/test-infra repository

Option 2 will be better for long term sustainability (my opinion). As the more jobs get added for each layered product, that will require numerous PR's into test-infra repository to include them in the dashboards.

In order to do this it would require a PR to be submitted to the `ci-tools` repository to add support for generating dashboards for layered product interop testing per openshift release. Depending on how we set it up in their program, we would not need to do anything. Just as long as the naming of the jobs are set up for the program to handle them. Example:

```
periodic-ci-lp-interop-quay-1.0-openshift-4.13
periodic-ci-quay-1.0-openshift-4.13-lp-interop
etc.. need to define what standard we want
```

We name all our jobs using a unique string such as `-lp-interop` or `-lp-interop-`, then as new jobs are added and the job is run on a daily basis, the dashboards will always be up to date. So we could end up having dashboards such as:

```
redhat-openshift-lp-interop-release-4.13-informing
redhat-openshift-lp-interop-release-4.14-informing
etc..
```

It would then pick up the OCP version from either the job name `-X.Y-` or it would get it from the jobs labels `job-release: "4.12"`.

We just would need to add a new case statement to their program: https://github.com/ryankwilliams/ci-tools/commit/1a562b995ac00b00d44f3b1ba8aa9e315bde47a0. If this is path we go, happy to submit this PR. This was done similar for the OSDE2E platform work, allowing for auto creating dashboards based on a label `osde2e` within the jobs list `openshift/release/core-services/testgrid-config-generator/_allow-list.yaml` . This is great if you have small amounts of jobs. Going this path, of just having a new label and adding the jobs ourself, still would require numerous PRs.

Interested to hear thoughts on this. This would allow for our own dashboards per OCP release. Then eventually when we are at a point and really want these to be informing the OCP releases, that can then be explored.

Author: @ryankwilliams I like this approach much more than manually creating the dashboards and doing separate PRs. Adam and I had discussed modifying the testgrid-config-generator, but weren't sure if a TestGrid PoC was worth the cycles just yet, but it looks like you've already got the solution, and it seems like we all agree on using TestGrid.

Personally, I like the `lp-interop` identifier. We will need a unique identifier for Report Portal reporting, if we use it, anyway. @ascerra are you okay using `lp-interop` as an identifier rather than `cspi-interop`? If so, I think it would be good to submit @ryankwilliams' PR. If that gets accepted, I can change this documentation to reflect the changes.

All that being said, @ryankwilliams, how do you feel about keeping documentation for creating/modifying a TestGrid dashboard, just in case? I can modify the documentation's location and wording to reflect that it wouldn't be a normal thing to do, but it is doable if needed.

Reviewer 2: @calebevans Sounds good, I can work on submitting a PR next week. I just will wait for @ascerra feedback here.

I prefer to use `lp-interop` over `cspi-interop` this way it is move specific to actually what is tested "layered products" where as `cspi` can mean a handful of items. Sticking with this model, could allow us to expand for add ons `add-on-interop` for integration testing between add ons (just as an example).

@calebevans That works for me, I am fine to leave it as a reference for everyone's awareness and once finalized, we just update it to include this new approach for how the dashboards are auto generated.

Reviewer 3: Automatic creation of dashboards sounds great, it makes our push to prod process even more lightweight.

`lp-interop` works for me.

Determining where we should add the unique identifier should probably be settled here as well

We could either add it in the variant (tail end of the filename) or in the [test stanza as: field ](https://github.com/openshift/release/blob/master/ci-operator/config/rhpit/interop-tests/rhpit-interop-tests-master__ocp-412-quay37.yaml#L30) 

I'm not sure adding it in both places will be ideal. I guess this also depends on the report portal identification of jobs as well.

My current vote would be to include `-lp-interop` at the tail end of the filename as the variant.

So using the quay config as an example
`config/rhpit/interop-tests/rhpit-interop-tests-master__ocp-412-quay37-lp-interop.yaml`

Author: Regarding where the unique identifier is:

My vote is also in the filename. The Report Portal automation I wrote uses the filename to find jobs on a unique identifier. Even if we don't end up using Report Portal, if we decided to write any other automation around the unique identifier, we don't want to load every YAML file, check for the value, then move on if it doesn't match. Much more resource friendly to use the filename.

I can start working on changing this documentation to focus on "if you need to create/modify a TestGrid dashboard manually" and write about how it should automatically be updated if the unique identifier is met. 

@ryankwilliams do you want to submit a PR to support the automatic creation of dashboards in TestGrid?

Reviewer 2: @calebevans @ascerra Yes, I can submit the PR tomorrow and link it here once opened.

As long as the string `-lp-interop` gets added within the prow job name, that is what is needed for the auto dashboard creation to work. 

I do think we should revisit the overall naming of these jobs and come up with a standard, so its clear and easy to follow.

Author: @ryankwilliams I tested that it would be in Prow job name if the file name contains that string earlier today, and it does show up in the Prow job name. 

I agree that we need a standard on naming, but I think we are pretty close, considering OpenShift CI forces us into the `{github org}-{github repo}-{branch}-****.yaml` format. We can talk about the naming standardization sometime this week.

Reviewer 2: @calebevans Thank you for confirming it shows up in the prow job name.

Sure, feel free to add me to a call or discussion about the naming standard. I will ping both you and @ascerra once I open the PR for auto-creating interop dashboards in test grid. I was unable to today but plan to do it tomorrow.

> TestGrid is a Kubernetes community project that allow users to create dashboards of Prow job results. TestGrid uses it's configuration files to build these dashboards, retrieve the Prow job results of all jobs defined in the dashboards, and displays the results in a grid pattern

Reviewer: ```suggestion
TestGrid is a Kubernetes community project that allow users to create dashboards of Prow job results. TestGrid uses its configuration files to build these dashboards, retrieve the Prow job results of all jobs defined in the dashboards, and display the results in a grid pattern.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

      - `url`: URL template. This should not change for our purposes.

- `test_groups`: A list of Prow jobs to be able to use in your dashboard.
  - `name`: Essentially an alias for the Prow job name. We use this name in the `test_group_name` key above. This value needs to be unique to not only this document, but other `test_groups` items in other config files.
  - `gcs_prefix`: The Google Cloud Storage prefix path for the Prow results. Should always be `origin-ci-test/logs/` plus the name of the Prow job.
  - `days_of_results`: How many days of results should be seen on TestGrid. I haven't seen this value go above 60 days, but I also haven't tried going any higher than that.

### How to Create a New Dashboard

Creating a new TestGrid dashboard is very simple. Before you get started you will need to [fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/about-forks) the [kubernetes/test-infra](https://github.com/kubernetes/test-infra) repository into your personal GitHub organization.

Once you have your fork created, follow these instructions to create a new dashboard:

1. Prepare Git:

>    1.  Clone a copy of your fork to your computer

Reviewer: ```suggestion
   1.  Clone a copy of your fork to your computer.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

   2.  Create a new branch and name it whatever you'd like.

> 2. In your new branch, create a new file in the `config/testgrids/openshift` directory. The name of the file should be the same name as your new dashboard

Reviewer: ```suggestion
2. In your new branch, create a new file in the `config/testgrids/openshift` directory. The name of the file should be the same name as your new dashboard.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

> **IMPORTANT**:
>
> Do not use `-release-` at all in your file name or your dashboard name. This will break the OpenShift CI automation that generates their TestGrid dashboards. See this pull request for more information: [kubernetes/test-infra PR #28593](https://github.com/kubernetes/test-infra/pull/28593 )

3. Write your configuration file.
   - **Note:** Use the example in the [TestGrid Configuration Example and Description](#testgrid-configuration-example-and-description) section of this document.
4. After you have pushed your changes to your fork, follow the [pull request process](#pull-request-process) to submit a PR and have your dashboard created.

### How to Add a Job to a Dashboard

If there is already a suitable dashboard created in TestGrid to hold your new Prow job, follow this guide to get it appended to the existing dashboard. Before you get started you will need to [fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/about-forks) the [kubernetes/test-infra](https://github.com/kubernetes/test-infra) repository into your personal GitHub organization.

1. Prepare Git:

>    1.  Clone a copy of your fork to your computer

Reviewer: ```suggestion
   1.  Clone a copy of your fork to your computer.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)

   2.  Create a new branch and name it whatever you'd like.
2.  In your new branch, find the configuration file for the dashboard you'd like to use.
    - **Note**: The config files should be in the `config/testgrids/openshift` directory and the name of the file should be the same as the name on the TestGrid website.
    - **Example:** `config/testgrids/openshift/redhat-openshift-4.12-interop.yaml`
3. Create an entry in the `test_groups` stanza of the config file.
    - **Note:** Use the example in the [TestGrid Configuration Example and Description](#testgrid-configuration-example-and-description) section of this document.

>     - **Note:** The `name` value in this item needs to be unique, not just in the document but in the other configs as well. You may want to search the repo for the string you decide on to make sure it isn't used.

Reviewer: ```suggestion
    - **Note:** The `name` value in this item needs to be unique, not just in the document, but in the other configs as well. You may want to search the repo for the string you decide on to make sure it isn't used.
```

Author: Resolved in [**39fe241**7dca0aa2df5e3e0cb4676f4fa7ac99bf7](https://github.com/CSPI-QE/ocp-ci-docs/pull/10/commits/39fe2417dca0aa2df5e3e0cb4676f4fa7ac99bf7)
